In [1]:
import pandas as pd

store = pd.read_csv('../data/SuperstoreSample.csv', encoding='cp1252',dtype={'Order ID': str, 'Customer ID': str, 'Postal Code': str})

# Displaying rows and columns 
print(f"Store Data CSV → Rows: {store.shape[0]}, Columns: {store.shape[1]}")
print(store.dtypes)


Store Data CSV → Rows: 9994, Columns: 21
Row ID             int64
Order ID             str
Order Date           str
Ship Date            str
Ship Mode            str
Customer ID          str
Customer Name        str
Segment              str
Country              str
City                 str
State                str
Postal Code          str
Region               str
Product ID           str
Category             str
Sub-Category         str
Product Name         str
Sales            float64
Quantity           int64
Discount         float64
Profit           float64
dtype: object


In [2]:
#Drop unnecessary columns
store = store.drop(columns=['Row ID'])
store = store.drop(columns=['Customer Name'])

1.'Row_ID' column was excluded from analysis because it does not provide analytical values, it serves only as a unique identifier.


2.'Customer Name' column was excluded becuase it does not provide analytical value and duplicates the information represented by Customer ID. 'Customer ID' column was kept for potential grouping and customer-level analysis.

In [3]:
# Converting Order Date to datetime format
store['Order Date'] = pd.to_datetime(store['Order Date'])

#Converting Ship Date to datetime format
store['Ship Date']  = pd.to_datetime(store['Ship Date'])

print(store.dtypes)

Order ID                   str
Order Date      datetime64[us]
Ship Date       datetime64[us]
Ship Mode                  str
Customer ID                str
Segment                    str
Country                    str
City                       str
State                      str
Postal Code                str
Region                     str
Product ID                 str
Category                   str
Sub-Category               str
Product Name               str
Sales                  float64
Quantity                 int64
Discount               float64
Profit                 float64
dtype: object


In [4]:
#Checking for null values
store.isnull().sum()

Order ID        0
Order Date      0
Ship Date       0
Ship Mode       0
Customer ID     0
Segment         0
Country         0
City            0
State           0
Postal Code     0
Region          0
Product ID      0
Category        0
Sub-Category    0
Product Name    0
Sales           0
Quantity        0
Discount        0
Profit          0
dtype: int64

In [5]:
#Checking for duplicates
store.duplicated().sum()

np.int64(1)

In [6]:
#Creating additional colums for time analysis 
store['Order Month'] = store['Order Date'].dt.month
store['Order Year'] = store['Order Date'].dt.year

#Calc shipping duration (days between order and delivery)
store['Shipping Duration'] = (store['Ship Date'] - store['Order Date']).dt.days

#Calc profit margin as %
store['Profit Margin'] =  (store['Profit'] / store['Sales']) * 100

#Categorising profit status as 'Profit' or 'Loss'
store['Profit Status'] = store['Profit'].apply(lambda x: 'Loss' if x < 0 else 'Profit')


print(store.columns)

Index(['Order ID', 'Order Date', 'Ship Date', 'Ship Mode', 'Customer ID',
       'Segment', 'Country', 'City', 'State', 'Postal Code', 'Region',
       'Product ID', 'Category', 'Sub-Category', 'Product Name', 'Sales',
       'Quantity', 'Discount', 'Profit', 'Order Month', 'Order Year',
       'Shipping Duration', 'Profit Margin', 'Profit Status'],
      dtype='str')


Several transactions show negative values within the 'Profit' column, indicating loss-making sales. As expected, these cases will result in negative 'Profit Margin(%)' values. Further analysis will focus on identifying drivers such as discounts or inefficient pricing.

In [7]:
#Validation
store.isnull().sum()

Order ID             0
Order Date           0
Ship Date            0
Ship Mode            0
Customer ID          0
Segment              0
Country              0
City                 0
State                0
Postal Code          0
Region               0
Product ID           0
Category             0
Sub-Category         0
Product Name         0
Sales                0
Quantity             0
Discount             0
Profit               0
Order Month          0
Order Year           0
Shipping Duration    0
Profit Margin        0
Profit Status        0
dtype: int64

In [8]:
#More Validations 
print(store['Profit Status'].value_counts())
print(f"Total Rows: {store.shape[0]}")
print(f"Rows with negative shipping duration: {store[store['Shipping Duration'] < 0].shape[0]}")

Profit Status
Profit    8123
Loss      1871
Name: count, dtype: int64
Total Rows: 9994
Rows with negative shipping duration: 0


In [9]:
store.head()

,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Segment,Country,City,State,Postal Code,...,Product Name,Sales,Quantity,Discount,Profit,Order Month,Order Year,Shipping Duration,Profit Margin,Profit Status
0,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Consumer,United States,Henderson,Kentucky,42420,...,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136,11,2016,3,16.00,Profit
1,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Consumer,United States,Henderson,Kentucky,42420,...,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820,11,2016,3,30.00,Profit
2,CA-2016-138688,2016-06-12,2016-06-16,Second Class,DV-13045,Corporate,United States,Los Angeles,California,90036,...,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714,6,2016,4,47.00,Profit
3,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Consumer,United States,Fort Lauderdale,Florida,33311,...,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310,10,2015,7,-40.00,Loss
4,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Consumer,United States,Fort Lauderdale,Florida,33311,...,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164,10,2015,7,11.25,Profit


In [10]:
#Additional column for discount percentage
store['Discount %'] = store['Discount'] * 100

In [11]:
#Validation
print(store[['Discount', 'Discount %']].head())

   Discount  Discount %
0      0.00         0.0
1      0.00         0.0
2      0.00         0.0
3      0.45        45.0
4      0.20        20.0


In [12]:
store[['Sales', 'Profit', 'Discount', 'Shipping Duration', 'Profit Margin', 'Discount %']].describe()

,Sales,Profit,Discount,Shipping Duration,Profit Margin,Discount %
count,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000
mean,229.858001,28.656896,0.156203,3.958175,12.031393,15.620272
std,623.245101,234.260108,0.206452,1.747567,46.675435,20.645197
min,0.444000,-6599.978000,0.000000,0.000000,-275.000000,0.000000
25%,17.280000,1.728750,0.000000,3.000000,7.500000,0.000000
50%,54.490000,8.666500,0.200000,4.000000,27.000000,20.000000
75%,209.940000,29.364000,0.200000,5.000000,36.250000,20.000000
max,22638.480000,8399.976000,0.800000,7.000000,50.000000,80.000000


1.Sales values are right-skewed, with a mean (229.86) larger than the median (54.49). This shows that most transactions are of low value, while a small number of high-value sales bring the overall sales average upward. This suggests the presence of outliers.


2.Profit values show large variability. Average profit is 28.66, the median is at 8.67, which shows that most transactions generate relatively small profits. Max(8399.976000) and Min(-6599.978000) values suggest inconsistent profitability and influence of extreme outliers.


3.Shipping duration looks relatively stable.


4.Discount % values - most transactions either have no discount or discount rate of around 20%, while a smaller number of transactions receive significantly higher discounts (may contribute to fluctuations in profitability)


5.Profit margins range from -275% to 50%, with an average of 12%. The minimum value (-275.000000) indicates the presence of extreme loss-making transactions. This suggests that certain transactions are heavily impacted by factors such as discounts or high costs.

In [13]:
#Calc unit price 
store['Unit Price'] = store['Sales'] / store['Quantity']
store['Unit Price'].describe()

count    9994.000000
mean       60.919569
std       142.927440
min         0.336000
25%         5.470000
50%        16.270000
75%        63.940000
max      3773.080000
Name: Unit Price, dtype: float64

'Unit Price' shows high variability, could possibly be a reflection of the differences across product sub-categories. This relationship will be explored further.

In [14]:
store.to_csv('../data/superstore_cleaned.csv', index=False)